cpp/src/arrow/c/dlpack.cc

```diff
@@ -0,0 +1,139 @@
+    default:
+      return Status::TypeError("DataType is not compatible with DLPack spec: ",
+                               type.ToString());
+  }
+}
+
+}  // namespace
+struct ManagerCtx {
+  std::shared_ptr<ArrayData> ref;
+  DLManagedTensor tensor;
+};
+
+Result<DLManagedTensor*> ExportArray(const std::shared_ptr<Array>& arr) {
+  // Define DLDevice struct nad check if array type is supported
+  // by the DLPack protocol at the same time. Raise TypeError if not.
+  // Supported data types: int, uint, float with no validity buffer.
+  ARROW_ASSIGN_OR_RAISE(auto device, ExportDevice(arr))
+
+  // Define the DLDataType struct
+  const DataType* arrow_type = arr->type().get();
+  ARROW_ASSIGN_OR_RAISE(auto dlpack_type, GetDLDataType(*arrow_type));
+
+  // Create ManagerCtx with the reference to
+  // the data of the array
+  std::shared_ptr<ArrayData> array_ref = arr->data();
+  std::unique_ptr<ManagerCtx> ctx(new ManagerCtx);
+  ctx->ref = array_ref;
+
+  // Define the data pointer to the DLTensor
+  // If array is of length 0, data pointer should be NULL
+  if (arr->length() == 0) {
+    ctx->tensor.dl_tensor.data = NULL;
+  } else if (arr->offset() > 0) {
+    const auto byte_width = arr->type()->byte_width();
+    const auto start = arr->offset() * byte_width;
+    ARROW_ASSIGN_OR_RAISE(auto sliced_buffer,
+                          SliceBufferSafe(array_ref->buffers[1], start));
+    ctx->tensor.dl_tensor.data =
+        const_cast<void*>(reinterpret_cast<const void*>(sliced_buffer->address()));
+  } else {
+    ctx->tensor.dl_tensor.data = const_cast<void*>(
+        reinterpret_cast<const void*>(array_ref->buffers[1]->address()));
+  }
+
+  ctx->tensor.dl_tensor.device = device;
+  ctx->tensor.dl_tensor.ndim = 1;
+  ctx->tensor.dl_tensor.dtype = dlpack_type;
+  ctx->tensor.dl_tensor.shape = const_cast<int64_t*>(&array_ref->length);
+  ctx->tensor.dl_tensor.strides = NULL;
+  ctx->tensor.dl_tensor.byte_offset = 0;
+
+  ctx->tensor.manager_ctx = ctx.get();
+  ctx->tensor.deleter = [](struct DLManagedTensor* self) {
+    delete reinterpret_cast<ManagerCtx*>(self->manager_ctx);
+  };
+  return &ctx.release()->tensor;
+}
+
+Result<DLDevice> ExportDevice(const std::shared_ptr<Array>& arr) {
+  // Check if array is supported by the DLPack protocol.
+  if (arr->null_count() > 0) {
+    return Status::TypeError("Can only use DLPack on arrays with no nulls.");
+  }
+  const DataType* arrow_type = arr->type().get();
+  if (arrow_type->id() == Type::BOOL) {
+    return Status::TypeError("Bit-packed boolean data type not supported by DLPack.");
+  }
+  if (!is_integer(arrow_type->id()) && !is_floating(arrow_type->id())) {
+    return Status::TypeError("DataType is not compatible with DLPack spec: ",
+                             arrow_type->ToString());
+  }
+
+  // Define DLDevice struct
+  DLDevice device;
+  if (arr->data()->buffers[1]->device_type() == DeviceAllocationType::kCPU) {
```

Comment:
I think it should be safe based on the type checks above, but just confirming if we had a type of `NA` that we'd error before hitting here, since I imagine that accessing `buffers[1]` wouldn't be safe on that type?

Reply:
Yes, that should be safe. There is a test in Python that checks that: https://github.com/apache/arrow/pull/38472/files#diff-1368d4aa013b54f08fa17e63822e01f6a879ae0ddafb3c1ba6e3e919750c5d7eR113

```python
    arr = pa.array([])
    with pytest.raises(TypeError, match="DataType is not compatible with DLPack spec"):
        np.from_dlpack(arr)
```

but I will add it to C++ also.